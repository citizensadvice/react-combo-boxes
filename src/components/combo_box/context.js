import { createContext } from 'react';
import { visuallyHiddenClassName } from '../../constants/visually_hidden_class_name';

export const Context = createContext({
  visuallyHiddenClassName,
  dispatch: null,
});
