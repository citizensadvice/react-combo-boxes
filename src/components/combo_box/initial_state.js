export function initialState({ selectedOption }) {
  return {
    inlineAutoselect: false,
    expanded: false,
    focusListBox: false,
    search: null,
    focusedOption: selectedOption,
    suggestedOption: null,
  };
}
