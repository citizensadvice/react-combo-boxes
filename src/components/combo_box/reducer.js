import { shallowEqualObjects } from 'shallow-equal';
import { SET_SEARCH, SET_EXPANDED, SET_CLOSED, SET_FOCUSED_OPTION, SET_FOCUS_LIST_BOX } from './actions';

// AT RISK: It is debatable autoselect in this form is actually useful
function applyAutoselect(state, { type, ...params }, props) {
  const { autoselect } = props;

  if (!autoselect) {
    return state;
  }

  switch (type) {
    case SET_FOCUSED_OPTION:
    case SET_SEARCH: {
      const { lastKeyRef: { current: key } } = props;
      const { focusListBox, search, inlineAutoselect, suggestedOption } = state;

      if (focusListBox || !search || !params.autoselect) {
        break;
      }

      if (type === SET_SEARCH && key === 'Backspace' && inlineAutoselect) {
        return {
          ...state,
          focusedOption: null,
          inlineAutoselect: false,
          search: search.slice(0, -1),
        };
      }

      if (key === 'Backspace' || key === 'Delete') {
        return {
          ...state,
          focusedOption: null,
          inlineAutoselect: false,
        };
      }

      if (!suggestedOption) {
        break;
      }

      const { inputRef: { current: { selectionStart } } } = props;

      return {
        ...state,
        focusedOption: suggestedOption,
        inlineAutoselect: autoselect === 'inline' && selectionStart === search.length,
      };
    }
    default:
  }

  return {
    ...state,
    inlineAutoselect: false,
  };
}

function applySuggestedOption(state, { type }, props) {
  switch (type) {
    case SET_FOCUSED_OPTION:
    case SET_SEARCH: {
      const { options, findSuggestion } = props;
      const { focusListBox, search } = state;

      if (!search || focusListBox) {
        return {
          ...state,
          suggestedOption: null,
        };
      }

      let suggestedOption = null;
      for (let i = 0; i < options.length; i += 1) {
        const result = findSuggestion(options[i], search);
        if (result) {
          suggestedOption = options[i];
          break;
        }
        if (result === false) {
          break;
        }
      }

      return {
        ...state,
        suggestedOption,
      };
    }
    default:
  }

  return state;
}

function reduce(state, { type, ...params }, props) {
  switch (type) {
    case SET_SEARCH: {
      const { search } = params;
      const { selectedOption } = props;
      let { focusedOption } = state;
      if (!search) {
        focusedOption = null;
      } else if (!state.expanded) {
        focusedOption = selectedOption;
      }
      return {
        ...state,
        search,
        focusedOption,
        expanded: true,
        focusListBox: false,
      };
    }
    case SET_EXPANDED: {
      const { selectedOption } = props;

      return {
        ...state,
        expanded: true,
        focusedOption: selectedOption,
        focusListBox: true,
      };
    }
    case SET_CLOSED: {
      const { expanded } = params;

      return {
        ...state,
        expanded: !!expanded,
        focusedOption: null,
        focusListBox: false,
        search: null,
      };
    }
    case SET_FOCUSED_OPTION: {
      const {
        focusListBox,
        focusedOption,
        expanded,
      } = params;

      return {
        ...state,
        expanded: expanded ?? true,
        focusListBox: (focusListBox == null ? state.focusListBox : focusListBox) && focusedOption,
        focusedOption,
      };
    }
    case SET_FOCUS_LIST_BOX: {
      const {
        focusListBox,
      } = params;

      return {
        ...state,
        focusListBox,
      };
    }
    /* istanbul ignore next */
    default:
      return state;
  }
}

export function reducer(state, action, props) {
  const newState = [
    reduce,
    applySuggestedOption,
    applyAutoselect,
  ].reduce((currentState, fn) => fn(currentState, action, props), state);

  if (shallowEqualObjects(newState, state)) {
    return state;
  }

  return newState;
}
