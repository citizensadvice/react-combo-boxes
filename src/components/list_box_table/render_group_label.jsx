import { makeBEMClass } from '../../helpers/make_bem_class';

export function renderGroupLabel(props, componentState, componentProps) {
  const {
    classPrefix,
    columns,
    renderTableGroupRow,
    renderTableGroupHeaderCell,
  } = componentProps;

  const {
    group: { key },
  } = componentState;

  return renderTableGroupRow(
    {
      className: makeBEMClass(classPrefix, 'table-group-row'),
      children: renderTableGroupHeaderCell(
        {
          ...props,
          colSpan: Object.keys(columns).length,
          className: makeBEMClass(classPrefix, 'table-group-header'),
          id: key,
        },
        componentState,
        componentProps,
      ),
    },
    componentState,
    componentProps,
  );
}
