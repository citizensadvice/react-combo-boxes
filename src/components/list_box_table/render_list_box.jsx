import React from 'react';
import { makeBEMClass } from '../../helpers/make_bem_class';

export function renderListBox({ children, hidden, ref, ...props }, componentState, componentProps) {
  const {
    classPrefix,
    columns,
    renderTableWrapper,
    renderTable,
    renderTableHeaderCell,
  } = componentProps;
  const hasHeader = columns.some(({ label }) => label);

  return renderTableWrapper({
    hidden,
    className: makeBEMClass(classPrefix, 'listbox', hasHeader && 'listbox--header'),
    onMouseDown: (e) => e.preventDefault(),
    ref,
    children: renderTable({
      role: 'listbox',
      ...props,
      className: makeBEMClass(classPrefix, 'table'),
      children: (
        <>
          <colgroup>
            {columns.map(({ name, colHtml }) => (
              <col key={name} {...colHtml} />
            ))}
          </colgroup>
          {hasHeader && (
            <thead role="presentation">
              <tr role="presentation">
                {columns.map((column) => renderTableHeaderCell({
                  'aria-hidden': 'true',
                  key: column.name,
                  className: makeBEMClass(classPrefix, 'table-header'),
                  children: column.label,
                }, { ...componentState, column }, componentProps))}
              </tr>
            </thead>
          )}
          <tbody role="presentation">
            {children}
          </tbody>
        </>
      ),
    }, componentState, componentProps),
  }, componentState, componentProps);
}
