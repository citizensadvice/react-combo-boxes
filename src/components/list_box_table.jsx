import React, { forwardRef, useContext, useMemo } from 'react';
import PropTypes from 'prop-types';
import { Context } from '../context';
import { renderGroupedOptions } from '../helpers/render_grouped_options';
import { classPrefix } from '../constants/class_prefix';

export const ListBoxTable = forwardRef(({ focusedRef, hidden, onSelectOption, ...props }, ref) => {
  const context = useContext(Context);
  const {
    currentOption,
    props: {
      columns: rawColumns,
      options,
      ListBoxListComponent = 'div',
      listBoxListProps,
      tableProps,
      tableHeaderProps,
      tableGroupRowProps,
      tableGroupHeaderProps,
      tableRowProps,
      tableCellProps,
      ValueComponent, valueProps,
      visuallyHiddenClassName,
    },
  } = context;

  const columns = useMemo(() => (
    rawColumns.map((column) => {
      if (typeof column === 'string') {
        return {
          name: column,
        };
      }
      return column;
    })
  ), [rawColumns]);

  return (
    <ListBoxListComponent
      hidden={hidden}
      className={`${classPrefix}listbox`}
      onMouseDown={(e) => e.preventDefault()}
      {...listBoxListProps}
    >
      <table
        ref={ref}
        role="listbox"
        {...props}
        className={`${classPrefix}listbox__table`}
        {...tableProps}
      >
        <colgroup>
          {columns.map(({ name, html }) => (
            <col key={name} {...html} />
          ))}
        </colgroup>
        {columns.some(({ label }) => label) && (
          <thead role="presentation">
            <tr>
              {columns.map(({ name, label }) => (
                <th
                  aria-hidden="true"
                  key={name}
                  className={`${classPrefix}listbox__table-header`}
                  {...tableHeaderProps}
                >
                  {label}
                </th>
              ))}
            </tr>
          </thead>
        )}
        <tbody role="presentation">
          {renderGroupedOptions({
            options,
            renderGroup(group) {
              const { key, html, label, children: groupChildren } = group;
              return (
                <Context.Provider
                  key={key}
                  value={{ ...context, group, columns }}
                >
                  <tr
                    className={`${classPrefix}listbox__table-group-row`}
                    {...tableGroupRowProps}
                  >
                    <th
                      aria-hidden="true"
                      colSpan={Object.keys(columns).length}
                      className={`${classPrefix}listbox__table-group-header`}
                      {...tableGroupHeaderProps}
                      {...html}
                    >
                      {label}
                    </th>
                  </tr>
                  {groupChildren}
                </Context.Provider>
              );
            },
            // eslint-disable-next-line react/prop-types
            renderOption(option) {
              const { key, html, disabled, group } = option;
              const selected = currentOption?.key === key;
              return (
                <Context.Provider
                  key={key}
                  value={{ ...context, selected, option, group, columns }}
                >
                  <tr
                    id={key}
                    role="option"
                    className={`${classPrefix}listbox__table-row`}
                    tabIndex={-1}
                    aria-selected={selected ? 'true' : null}
                    aria-disabled={disabled ? 'true' : null}
                    ref={selected ? focusedRef : null}
                    {...tableRowProps}
                    {...html}
                    onClick={disabled ? null : (e) => onSelectOption(e, option)}
                  >
                    {columns.map((column, index) => (
                      <Context.Provider
                        key={column.name}
                        value={{ ...context, selected, option, group, columns, column }}
                      >
                        <td
                          role="presentation"
                          className={`${classPrefix}listbox__table-cell`}
                          {...tableCellProps}
                        >
                          {group && index === 0 && (
                            <div className={visuallyHiddenClassName}>
                              {group.label}
                            </div>
                          )}
                          {column.label && (
                            <div className={visuallyHiddenClassName}>
                              {column.label}
                            </div>
                          )}
                          <ValueComponent
                            {...valueProps}
                          >
                            {option.value[column.name]}
                          </ValueComponent>
                        </td>
                      </Context.Provider>
                    ))}
                  </tr>
                </Context.Provider>
              );
            },
          })}
        </tbody>
      </table>
    </ListBoxListComponent>
  );
});

ListBoxTable.propTypes = {
  'aria-activedescendant': PropTypes.string,
  focusedRef: PropTypes.shape({
    current: PropTypes.instanceOf(Element),
  }),
  hidden: PropTypes.bool,
  id: PropTypes.string.isRequired,
  onSelectOption: PropTypes.func.isRequired,
};

ListBoxTable.defaultProps = {
  focusedRef: null,
  hidden: false,
  'aria-activedescendant': null,
};

ListBoxTable.displayName = 'ListBoxTable';
