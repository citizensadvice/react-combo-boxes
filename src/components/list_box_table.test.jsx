import React, { useState, useContext, forwardRef } from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ComboBox } from './combo_box';
import { ListBoxTable } from './list_box_table';
import { Context } from '../context';

const ComboBoxWrapper = forwardRef(({ value: initialValue, ...props }, ref) => {
  const [value, onValue] = useState(initialValue);
  return (
    <ComboBox id="id" value={value} onValue={onValue} {...props} ref={ref} />
  );
});

describe('used with a combo box', () => {
  describe('columns as names only', () => {
    const options = [
      { label: 'Apple', type: 'Fruit', colour: 'Green' },
      { label: 'Banana', type: 'Fruit', colour: 'Yellow' },
      { label: 'Potato', type: 'Vegetable', colour: 'Brown' },
      { label: 'Tomato', type: 'Vegetable', colour: 'Red', disabled: true },
    ];

    const columns = ['label', 'type'];

    it('renders a table as the list box', () => {
      const { container } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
        />
      ));
      expect(container).toMatchSnapshot();
    });

    it('allows selection by click', () => {
      const spy = jest.fn();
      const { getByRole, getAllByRole } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
          onValue={spy}
        />
      ));
      getByRole('combobox').focus();
      userEvent.click(getAllByRole('option')[1]);
      expect(spy).toHaveBeenCalledWith({ label: 'Banana', type: 'Fruit', colour: 'Yellow' });
    });

    it('allows selection by keyboard', () => {
      const spy = jest.fn();
      const { getByRole } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
          onValue={spy}
        />
      ));
      getByRole('combobox').focus();
      fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
      fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
      fireEvent.keyDown(document.activeElement, { key: 'Enter' });

      expect(spy).toHaveBeenCalledWith({ label: 'Banana', type: 'Fruit', colour: 'Yellow' });
    });
  });

  describe('columns with headers', () => {
    const options = [
      { label: 'Apple', type: 'Fruit', colour: 'Green' },
      { label: 'Banana', type: 'Fruit', colour: 'Yellow' },
      { label: 'Potato', type: 'Vegetable', colour: 'Brown' },
    ];

    const columns = [
      { name: 'label', label: 'Name' },
      { name: 'type', label: 'Type' },
    ];

    it('renders a table with headers', () => {
      const { container } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
        />
      ));
      expect(container).toMatchSnapshot();
    });

    it('allows selection by click', () => {
      const spy = jest.fn();
      const { getByRole, getAllByRole } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
          onValue={spy}
        />
      ));
      getByRole('combobox').focus();
      userEvent.click(getAllByRole('option')[1]);
      expect(spy).toHaveBeenCalledWith({ label: 'Banana', type: 'Fruit', colour: 'Yellow' });
    });

    it('allows selection by keyboard', () => {
      const spy = jest.fn();
      const { getByRole } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
          onValue={spy}
        />
      ));
      getByRole('combobox').focus();
      fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
      fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
      fireEvent.keyDown(document.activeElement, { key: 'Enter' });

      expect(spy).toHaveBeenCalledWith({ label: 'Banana', type: 'Fruit', colour: 'Yellow' });
    });
  });

  describe('columns with html', () => {
    const options = [
      { label: 'Apple', type: 'Fruit', colour: 'Green' },
      { label: 'Banana', type: 'Fruit', colour: 'Yellow' },
      { label: 'Potato', type: 'Vegetable', colour: 'Brown' },
    ];

    const columns = [
      { name: 'label', html: { className: 'foo' } },
      { name: 'type', html: { className: 'bar' } },
    ];

    it('renders a table with colgroups', () => {
      const { container } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
        />
      ));
      expect(container).toMatchSnapshot();
    });
  });

  describe('grouped', () => {
    const options = [
      { name: 'Apple', type: 'Fruit', colour: 'Green' },
      { name: 'Banana', type: 'Fruit', colour: 'Yellow' },
      { name: 'Potato', type: 'Vegetable', colour: 'Brown' },
    ];

    const columns = ['name', 'colour'];

    function map({ name, type }) {
      return {
        label: name,
        group: type,
      };
    }

    it('renders a table with groups', () => {
      const { container } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
          mapOption={map}
        />
      ));
      expect(container).toMatchSnapshot();
    });

    it('allows selection by click', () => {
      const spy = jest.fn();
      const { getByRole, getAllByRole } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
          onValue={spy}
          mapOption={map}
        />
      ));
      getByRole('combobox').focus();
      userEvent.click(getAllByRole('option')[1]);
      expect(spy).toHaveBeenCalledWith({ name: 'Banana', type: 'Fruit', colour: 'Yellow' });
    });

    it('allows selection by keyboard', () => {
      const spy = jest.fn();
      const { getByRole } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
          onValue={spy}
          mapOption={map}
        />
      ));
      getByRole('combobox').focus();
      fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
      fireEvent.keyDown(document.activeElement, { key: 'ArrowDown' });
      fireEvent.keyDown(document.activeElement, { key: 'Enter' });

      expect(spy).toHaveBeenCalledWith({ name: 'Banana', type: 'Fruit', colour: 'Yellow' });
    });
  });

  describe('customisation', () => {
    const options = [
      { name: 'Apple', type: 'Fruit', colour: 'Green' },
      { name: 'Banana', type: 'Fruit', colour: 'Yellow' },
      { name: 'Potato', type: 'Vegetable', colour: 'Brown' },
    ];

    const columns = [
      { name: 'name', label: 'Name' },
      { name: 'colour', label: 'Colour' },
    ];

    function map({ name, type }) {
      return {
        label: name,
        group: type,
      };
    }

    it('adds props to the components', () => {
      const { container } = render((
        <ComboBoxWrapper
          options={options}
          columns={columns}
          ListBoxComponent={ListBoxTable}
          mapOption={map}
          listBoxListProps={{ className: 'listbox' }}
          tableProps={{ className: 'table' }}
          tableHeaderProps={{ className: 'table-header' }}
          tableGroupRowProps={{ className: 'table-group-row' }}
          tableGroupHeaderProps={{ className: 'table-group-header' }}
          tableRowProps={{ className: 'table-row' }}
          tableCellProps={{ className: 'table-cell' }}
        />
      ));

      expect(container.querySelector('table').parentNode).toHaveClass('listbox');
      expect(container.querySelector('table')).toHaveClass('table');
      expect(container.querySelector('th')).toHaveClass('table-header');
      expect(container.querySelector('tbody tr:first-child')).toHaveClass('table-group-row');
      expect(container.querySelector('th[colspan]')).toHaveClass('table-group-header');
      expect(container.querySelector('tbody tr:nth-child(2)')).toHaveClass('table-row');
      expect(container.querySelector('td')).toHaveClass('table-cell');
    });

    describe('ListBoxListComponent', () => {
      it('allows the list box to be replaced', () => {
        const { container } = render((
          <ComboBoxWrapper
            options={options}
            columns={columns}
            ListBoxComponent={ListBoxTable}
            mapOption={map}
            ListBoxListComponent="dl"
          />
        ));

        expect(container.querySelector('table').parentNode.tagName).toEqual('DL');
      });
    });

    describe('visuallyHiddenClassName', () => {
      it('allows custom props', () => {
        const { getAllByRole, getByRole } = render((
          <ComboBoxWrapper
            options={options}
            columns={columns}
            ListBoxComponent={ListBoxTable}
            mapOption={map}
            visuallyHiddenClassName="bar"
          />
        ));
        getByRole('combobox').focus();
        expect(getAllByRole('option')[0].firstChild.firstChild).toHaveClass('bar');
      });
    });

    describe('ValueComponent', () => {
      it('allows the component to be replaced', () => {
        const { getByRole, container } = render((
          <ComboBoxWrapper
            options={options}
            columns={columns}
            ListBoxComponent={ListBoxTable}
            mapOption={map}
            ValueComponent="dl"
          />
        ));
        getByRole('combobox').focus();
        expect(container.querySelector('td > dl')).toHaveTextContent('Apple');
      });

      it('gives access to the context', () => {
        const spy = jest.fn();

        const ValueComponent = forwardRef((props, _) => {
          const context = useContext(Context);
          spy(context);
          return (
            <div {...props} />
          );
        });

        render((
          <ComboBoxWrapper
            options={options}
            columns={columns}
            ListBoxComponent={ListBoxTable}
            mapOption={map}
            ValueComponent={ValueComponent}
            foo="bar"
          />
        ));

        expect(spy).toHaveBeenCalledWith({
          'aria-autocomplete': 'none',
          'aria-busy': false,
          expanded: false,
          notFound: false,
          currentOption: null,
          search: null,
          suggestedOption: null,
          selected: false,
          props: expect.objectContaining({
            foo: 'bar',
            options: expect.any(Array),
            value: null,
          }),
          group: expect.objectContaining({
            label: 'Fruit',
            options: expect.any(Array),
          }),
          option: expect.objectContaining({
            label: 'Apple',
            group: expect.objectContaining({
              label: 'Fruit',
            }),
          }),
          columns: [
            { label: 'Name', name: 'name' },
            { label: 'Colour', name: 'colour' },
          ],
          column: {
            label: 'Name', name: 'name',
          },
        });
      });
    });

    describe('valueProps', () => {
      it('allows the props to be amended', () => {
        const { getByRole, container } = render((
          <ComboBoxWrapper
            options={options}
            columns={columns}
            ListBoxComponent={ListBoxTable}
            mapOption={map}
            ValueComponent="dl"
            valueProps={{ className: 'foo' }}
          />
        ));
        getByRole('combobox').focus();
        expect(container.querySelector('td > dl')).toHaveClass('foo');
      });
    });
  });
});
