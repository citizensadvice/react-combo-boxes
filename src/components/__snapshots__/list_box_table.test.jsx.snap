// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`used with a combo box columns as names only renders a table as the list box 1`] = `
<div>
  <div
    aria-busy="false"
    class="react-combo-boxes-combobox"
  >
    <input
      aria-autocomplete="none"
      aria-controls="id_listbox"
      aria-describedby="id_found_description"
      aria-expanded="false"
      autocomplete="off"
      class="react-combo-boxes-combobox__input"
      id="id"
      role="combobox"
      type="text"
      value=""
    />
    <span
      aria-hidden="true"
      class="react-combo-boxes-combobox__open-button"
      id="id_open_button"
    />
    <span
      aria-hidden="true"
      class="react-combo-boxes-combobox__clear-button"
      hidden=""
      id="id_clear_button"
    />
    <div
      class="react-combo-boxes-listbox"
      hidden=""
    >
      <table
        class="react-combo-boxes-listbox__table"
        id="id_listbox"
        role="listbox"
        tabindex="-1"
      >
        <colgroup>
          <col />
          <col />
        </colgroup>
        <tbody
          role="presentation"
        >
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_apple"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Apple
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Fruit
            </td>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_banana"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Banana
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Fruit
            </td>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_potato"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Potato
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Vegetable
            </td>
          </tr>
          <tr
            aria-disabled="true"
            class="react-combo-boxes-listbox__table-row"
            id="id_option_tomato"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Tomato
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Vegetable
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_found_description"
    />
    <div
      class="react-combo-boxes-combobox__not-found"
      hidden=""
      id="id_not_found"
    />
  </div>
</div>
`;

exports[`used with a combo box columns with headers renders a table with headers 1`] = `
<div>
  <div
    aria-busy="false"
    class="react-combo-boxes-combobox"
  >
    <input
      aria-autocomplete="none"
      aria-controls="id_listbox"
      aria-describedby="id_found_description"
      aria-expanded="false"
      autocomplete="off"
      class="react-combo-boxes-combobox__input"
      id="id"
      role="combobox"
      type="text"
      value=""
    />
    <span
      aria-hidden="true"
      class="react-combo-boxes-combobox__open-button"
      id="id_open_button"
    />
    <span
      aria-hidden="true"
      class="react-combo-boxes-combobox__clear-button"
      hidden=""
      id="id_clear_button"
    />
    <div
      class="react-combo-boxes-listbox"
      hidden=""
    >
      <table
        class="react-combo-boxes-listbox__table"
        id="id_listbox"
        role="listbox"
        tabindex="-1"
      >
        <colgroup>
          <col />
          <col />
        </colgroup>
        <thead
          role="presentation"
        >
          <tr>
            <th
              aria-hidden="true"
              class="react-combo-boxes-listbox__table-header"
            >
              Name
            </th>
            <th
              aria-hidden="true"
              class="react-combo-boxes-listbox__table-header"
            >
              Type
            </th>
          </tr>
        </thead>
        <tbody
          role="presentation"
        >
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_apple"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Name
              </div>
              Apple
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Type
              </div>
              Fruit
            </td>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_banana"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Name
              </div>
              Banana
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Type
              </div>
              Fruit
            </td>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_potato"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Name
              </div>
              Potato
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Type
              </div>
              Vegetable
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_found_description"
    />
    <div
      class="react-combo-boxes-combobox__not-found"
      hidden=""
      id="id_not_found"
    />
  </div>
</div>
`;

exports[`used with a combo box columns with html renders a table with colgroups 1`] = `
<div>
  <div
    aria-busy="false"
    class="react-combo-boxes-combobox"
  >
    <input
      aria-autocomplete="none"
      aria-controls="id_listbox"
      aria-describedby="id_found_description"
      aria-expanded="false"
      autocomplete="off"
      class="react-combo-boxes-combobox__input"
      id="id"
      role="combobox"
      type="text"
      value=""
    />
    <span
      aria-hidden="true"
      class="react-combo-boxes-combobox__open-button"
      id="id_open_button"
    />
    <span
      aria-hidden="true"
      class="react-combo-boxes-combobox__clear-button"
      hidden=""
      id="id_clear_button"
    />
    <div
      class="react-combo-boxes-listbox"
      hidden=""
    >
      <table
        class="react-combo-boxes-listbox__table"
        id="id_listbox"
        role="listbox"
        tabindex="-1"
      >
        <colgroup>
          <col
            class="foo"
          />
          <col
            class="bar"
          />
        </colgroup>
        <tbody
          role="presentation"
        >
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_apple"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Apple
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Fruit
            </td>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_banana"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Banana
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Fruit
            </td>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_potato"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Potato
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Vegetable
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_found_description"
    />
    <div
      class="react-combo-boxes-combobox__not-found"
      hidden=""
      id="id_not_found"
    />
  </div>
</div>
`;

exports[`used with a combo box grouped renders a table with groups 1`] = `
<div>
  <div
    aria-busy="false"
    class="react-combo-boxes-combobox"
  >
    <input
      aria-autocomplete="none"
      aria-controls="id_listbox"
      aria-describedby="id_found_description"
      aria-expanded="false"
      autocomplete="off"
      class="react-combo-boxes-combobox__input"
      id="id"
      role="combobox"
      type="text"
      value=""
    />
    <span
      aria-hidden="true"
      class="react-combo-boxes-combobox__open-button"
      id="id_open_button"
    />
    <span
      aria-hidden="true"
      class="react-combo-boxes-combobox__clear-button"
      hidden=""
      id="id_clear_button"
    />
    <div
      class="react-combo-boxes-listbox"
      hidden=""
    >
      <table
        class="react-combo-boxes-listbox__table"
        id="id_listbox"
        role="listbox"
        tabindex="-1"
      >
        <colgroup>
          <col />
          <col />
        </colgroup>
        <tbody
          role="presentation"
        >
          <tr
            class="react-combo-boxes-listbox__table-group-row"
          >
            <th
              aria-hidden="true"
              class="react-combo-boxes-listbox__table-group-header"
              colspan="2"
            >
              Fruit
            </th>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_apple"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Fruit
              </div>
              Apple
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Green
            </td>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_banana"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Fruit
              </div>
              Banana
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Yellow
            </td>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-group-row"
          >
            <th
              aria-hidden="true"
              class="react-combo-boxes-listbox__table-group-header"
              colspan="2"
            >
              Vegetable
            </th>
          </tr>
          <tr
            class="react-combo-boxes-listbox__table-row"
            id="id_option_potato"
            role="option"
            tabindex="-1"
          >
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              <div
                class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
              >
                Vegetable
              </div>
              Potato
            </td>
            <td
              class="react-combo-boxes-listbox__table-cell"
              role="presentation"
            >
              Brown
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="visually-hidden visuallyhidden sr-only react-combo-boxes-sr-only"
      id="id_found_description"
    />
    <div
      class="react-combo-boxes-combobox__not-found"
      hidden=""
      id="id_not_found"
    />
  </div>
</div>
`;
